spine/model/layer/cnn/ppn.py

```diff
@@ -484,9 +492,75 @@ def process_loss_config(self, mask_loss='CE', resolution=5.,
         else:
             raise ValueError(
                     f"Mask loss name not recognized: {mask_loss}")
+            
+            
+    @staticmethod
+    def get_ppn_positives(coords : torch.Tensor, 
+                          labels: torch.Tensor, 
+                          ppn_labels: torch.Tensor, 
+                          resolution: float,
+                          offset: int):
+        """Get ppn positive label mask, but with cluster restrictions.
+
+        Parameters
+        ----------
+        coords : torch.Tensor
+            N x 3 3D coordinates of the image voxels
+        labels : torch.Tensor
+            N x 1 tensor of the particle id label for each voxel
+        ppn_labels : torch.Tensor
+            N_p x 6 PPN label points 
+        resolution : float
+            Distance from a label point in pixels within which a voxel is
+            
+        Warning
+        -------
+        If a given cluster has two ppn label points (tracks), then they must
+        appear consecutively in the ppn_labels tensor.
+
+        Returns
+        -------
+        positives : torch.Tensor
+            N x 1 tensor of the positive label mask
+        """
+        positives = torch.zeros(coords.shape[0], device=coords.device, dtype=torch.bool)
+        closests = -torch.ones(coords.shape[0], device=coords.device, dtype=torch.long)
+        
+        # test_mask = torch.zeros(coords.shape[0], device=coords.device, dtype=torch.bool)
+        
+        # print("Labels = ", labels.unique(return_counts=True), ppn_labels[:, PPN_LPART_COL])
+        
+        frag_ids = set()
+        
+        for i, pt in enumerate(ppn_labels):
+            
+            frag_id = int(pt[PPN_LPART_COL])
+            if frag_id in frag_ids:
+                continue
+            frag_ids.add(frag_id)
+            
+            mask = labels == frag_id
+            # test_mask[mask] = True
+            label_points = ppn_labels[ppn_labels[:, 5] == frag_id][:, COORD_COLS]
```

Comment:
If the comment above stand, no need to play this game, just compute the distance to the `pt` in the loop to the points within the corresponding particle. Keep the `mask`, of course.